wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceInfoViewModel.swift

```diff
@@ -169,52 +144,101 @@ final class DeviceInfoViewModel: ObservableObject {
     static func map(
         userClient: UserClient,
+        title: String,
+        addedDate: String,
+        proteusID: String?,
         isSelfClient: Bool,
         userSession: UserSession,
         credentials: ZMEmailCredentials?,
-        getUserClientFingerprintUseCase: GetUserClientFingerprintUseCaseProtocol,
-        e2eIdentityProvider: E2eIdentityProviding,
-        mlsProvider: MLSProviding
+        gracePeriod: TimeInterval,
+        conversationId: Data?,
+        getE2eIdentityEnabled: GetIsE2EIdentityEnabledUsecaseProtocol,
+        getE2eIdentityCertificates: GetE2eIdentityCertificatesUsecaseProtocol
     ) -> DeviceInfoViewModel {
         return DeviceInfoViewModel(
-            title: userClient.model ?? "",
-            addedDate: userClient.activationDate?.formattedDate ?? "",
-            proteusID: userClient.proteusSessionID?.clientID.fingerprintStringWithSpaces.uppercased() ?? "",
+            title: title,
+            addedDate: addedDate,
+            proteusID: proteusID?.uppercased().fingerprintStringWithSpaces ?? "",
             isProteusVerificationEnabled: userClient.verified,
             actionsHandler: DeviceDetailsViewActionsHandler(
                 userClient: userClient,
                 userSession: userSession,
                 credentials: credentials,
-                e2eIdentityProvider: e2eIdentityProvider,
-                mlsProvider: mlsProvider,
-                saveFileManager: SaveFileManager(systemFileSavePresenter: SystemSavePresenter())
+                conversationId: conversationId,
+                saveFileManager: SaveFileManager(systemFileSavePresenter: SystemSavePresenter()),
+                mlsClientResolver: MLSClientResolver(),
+                getE2eIdentityEnabled: getE2eIdentityEnabled,
+                getE2eIdentityCertificates: getE2eIdentityCertificates
             ),
-            userSession: userSession,
-            getUserClientFingerprint: getUserClientFingerprintUseCase,
             userClient: userClient,
-            isSelfClient: isSelfClient
+            isSelfClient: isSelfClient,
+            gracePeriod: gracePeriod
         )
     }
 }
+
+extension E2eIdentityCertificate {
+
+    // current default days the certificate is retained on server
+    private var kServerRetainedDays: Double { 28 * 24 * 60 * 60 }
+
+    // Randomising time so that not all clients update certificate at the same time
+    private var kRandomInterval: Double { Double(Int.random(in: 0..<86400)) }
+
+    var isExpired: Bool {
+        return expiryDate > Date.now
+    }
+
+    var isValid: Bool {
+        status == .valid
+    }
+
+    var isActivated: Bool {
+        return notValidBefore <= Date.now
+    }
+
+    var lastUpdatedDate: Date {
+        return notValidBefore + kServerRetainedDays + kRandomInterval
+    }
+
+    func shouldCertificateBeUpdated(with gracePeriod: TimeInterval) -> Bool {
+        guard isActivated,
+              isExpired,
```

Comment:
**question:**  should we update when the certificate is Expired?

Reply:
I guess so. Only when the certificate is revoked., the certificate should not be updated.